src/vs/workbench/contrib/notebook/browser/viewParts/notebookEditorToolbar.ts

```diff
@@ -1174,3 +617,105 @@ export function workbenchCalculateActions(initialPrimaryActions: IActionModel[],
+		});
+		return {
+			primaryActions: initialPrimaryActions.map(action => action.action),
+			secondaryActions: initialSecondaryActions
+		};
+	}
+
+	// step 2: check if they fit without labels
+	if ((visibleActionLength * ICON_ONLY_ACTION_WIDTH + (visibleActionLength - 1) * ACTION_PADDING) > leftToolbarContainerMaxWidth) {
+		return calculateIconOverflow(initialPrimaryActions, initialSecondaryActions, leftToolbarContainerMaxWidth);
+	}
+
+	// step 3: render as many actions as possible with labels, rest without.
+	let sum = 0;
+	let lastActionWithLabel = -1;
+	for (let i = 0; i < initialPrimaryActions.length; i++) {
+		sum += initialPrimaryActions[i].size + ACTION_PADDING;
+
+		if (initialPrimaryActions[i].action instanceof Separator) {
+			// find group separator
+			const remainingItems = initialPrimaryActions.slice(i + 1);
+			const newTotalSum = sum + (remainingItems.length === 0 ? 0 : (remainingItems.length * ICON_ONLY_ACTION_WIDTH + (remainingItems.length - 1) * ACTION_PADDING));
+			if (newTotalSum <= leftToolbarContainerMaxWidth) {
+				lastActionWithLabel = i;
+			}
+		} else {
+			continue;
+		}
+	}
+
+	// icons only don't fit either
+	// todo: don't quite understand this bit
+	if (lastActionWithLabel < 0) {
+		return calculateIconOverflow(initialPrimaryActions, initialSecondaryActions, leftToolbarContainerMaxWidth);
+	}
+
+
+	initialPrimaryActions.slice(0, lastActionWithLabel + 1).forEach(action => { action.renderLabel = true; });
+	initialPrimaryActions.slice(lastActionWithLabel + 1).forEach(action => { action.renderLabel = false; });
+	return {
+		primaryActions: initialPrimaryActions.map(action => action.action),
+		secondaryActions: initialSecondaryActions
+	};
+}
+
+function calculateIconOverflow(initialPrimaryActions: any[], initialSecondaryActions: any[], leftToolbarContainerMaxWidth: number) {
+	initialPrimaryActions.forEach(action => { action.renderLabel = false; });
+	const renderActions: IActionModel[] = [];
+	const overflow: IAction[] = [];
+
+	let currentSize = 0;
+	for (let i = 0; i < initialPrimaryActions.length; i++) {
+		const actionModel = initialPrimaryActions[i];
+		const itemSize = ICON_ONLY_ACTION_WIDTH;
+
+		if (currentSize + itemSize <= leftToolbarContainerMaxWidth) {
+			currentSize += ACTION_PADDING + itemSize;
+			renderActions.push(actionModel);
+		} else {
+			overflow.push(actionModel.action);
+		}
+	}
+
+	for (let i = (renderActions.length - 1); i !== 0; i--) {
+		const temp = renderActions[i];
+		if (temp.size === 0) {
+			continue;
+		}
+		if (temp.action instanceof Separator) {
+			renderActions.splice(i, 1);
+		}
+		break;
+	}
+
+
+	if (renderActions.length && renderActions[renderActions.length - 1].action instanceof Separator) {
+		renderActions.pop();
+	}
+
+	if (overflow.length !== 0) {
+		overflow.push(new Separator());
+	}
+
+
+	return {
+		primaryActions: renderActions.map(action => action.action),
+		secondaryActions: [...overflow, ...initialSecondaryActions]
+	};
```

Comment:
this function is very close to https://github.com/microsoft/vscode/blob/80ffe8ebdcb9828e59bea511ed39164026bab0f4/src/vs/workbench/contrib/notebook/browser/viewParts/notebookEditorToolbar.ts#L572-L591 and the only difference is all actions are labeless, let's unify them.